README.md

```diff
@@ -345,24 +345,73 @@ explained in terms of the sequential model. This is useful for trying to find
+         times with the part of the generated list that doesn't belong to the
+         prefix.
 
-  3. execute the prefix sequentially;
+  2. execute the prefix sequentially as in the section above (checking pre and
+     post-conditions);
 
-  4. execute the suffixes in parallel and gather the a trace (or history) of
-     invocations and responses of each action;
+  3. execute the suffixes in parallel without checking pre/post-conditions
+     and gather the trace (or history) of invocations and responses of each
+     action;
 
-  5. try to find a possible sequential interleaving of action invocations and
-     responses that respects the post-conditions.
+```
+                            ┌── no checks aside from ensuring no exception was thrown
+                            │
+                  ╭─────────┴──────────╮
+
+                  ┌─ [C] ──┐  ┌ [F, G] ┐   ◀─╮
+Commands: [A, B] ─┤        ├──┤        │     ├─ executed `concurrently`
+                  └ [D, E] ┘  └ [H, I] ┘   ◀─╯
+
+          ╰─┬──╯       ▲          ▲
+            │          ╰─────┬────╯
+            │                └── groups are not run in parallel
+            │                    i.e [C, D, E] will run (and
+            │                    finish) before F or H are
+            │                    started
+            │
+            └── pre/postconditions and invariant checked
+                executed sequentially
+```
+
+  4. if something goes wrong when executing the commands, shrink the generated
+     commands and present a minimal counterexample;
+
+  5. otherwise, try to find a possible sequential interleaving of action
+     invocations and responses that respects the post-conditions. For each
+     interleaving, this is done by advancing the `Concrete` model (starting at
+     `initalModel`) through the sequence of pairs of invocations to `command
+     Concrete` and the returned `response Concrete` emitted at step 3, check the
+     postcondition at each pair.
+
+  6. if no possible sequential interleaving was found, then shrink the generated
+     commands and present a minimal counterexample.
 
 The last step basically tries to find
 a [linearisation](https://en.wikipedia.org/wiki/Linearizability) of calls that
 could have happend on a single thread.
 
+Notice that step 5 above introduces a subtlety in the post-condition checks and
+transitions for the `model Concrete`. Despite the system under test running in a
+concurrent way, the model can still be designed to work sequentially as *it will
+not be run in parallel*, it will only be advanced in a sequential way when
+evaluating possible interleavings.
+
+As we cannot control the actual scheduling of the tasks, the sequence of
+commands (already fixed in a concrete prefix and a concrete list of suffixes) is
+actually executed several times by default, i.e. steps 2 to 6 will be executed
+multiple times for the same test case expecting that the scheduling of events
+varies between runs.
```

Comment:
Note that one can further introduce entropy here by introducing random sleeps in the semantic function.

Reply:
I added this to the end of the paragraph, I think it is an important subtlety worth mentioning